ta/remoteproc/remoteproc_core.c

```diff
@@ -0,0 +1,1212 @@
+					  PTA_RPROC_TLV_PARAM,
+					  exp_pt, params, NULL);
+		if (res)
+			return res;
+
+		p_tlv += U64_ALIGN_SZ(sizeof(struct remoteproc_tlv) +
+				      tlv_length);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result remoteproc_verify_firmware(struct remoteproc_context *ctx,
+					     uint8_t *fw_orig,
+					     uint32_t fw_orig_size)
+{
+	struct remoteproc_fw_hdr *hdr = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = get_rproc_pta_capabilities(ctx);
+	if (res)
+		return res;
+
+	/* Secure the firmware image depending on strategy */
+	if (!(ctx->hw_img_prot & PTA_RPROC_FW_WITH_HASH_TABLE) ||
+	    ctx->hw_fmt != PTA_RPROC_ELF_FMT) {
+		/*
+		 * Only hash table for ELF format support implemented
+		 * in a first step.
+		 */
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	res = remoteproc_save_fw_header(ctx, fw_orig, fw_orig_size);
+	if (res)
+		return res;
+
+	res = remoteproc_verify_header(ctx, fw_orig_size);
+	if (res)
+		goto free_hdr;
+
+	res = remoteproc_save_fw_trailer(ctx, fw_orig, fw_orig_size);
+	if (res)
+		goto free_hdr;
+
+	res = remoteproc_verify_trailer(ctx);
+	if (res)
+		goto free_tlr;
+
+	res = remoteproc_verify_signature(ctx);
+	if (res)
+		goto free_tlr;
+
+	res = remoteproc_set_platform_tlv(ctx);
+	if (res)
+		goto free_tlr;
+
+	/* Store location of the loadable binary in non-secure memory */
+	hdr = ctx->hdr;
+	ctx->fw_img_sz = hdr->img_len;
+	ctx->fw_img = fw_orig + sizeof(*hdr) + hdr->sign_len;
+
+	DMSG("Firmware image addr: %p size: %zu", ctx->fw_img,
+	     ctx->fw_img_sz);
+
+	return res;
+
+free_tlr:
+	TEE_Free(ctx->tlr);
+free_hdr:
+	TEE_Free(ctx->hdr);
+
+	return res;
+}
+
+static paddr_t remoteproc_da_to_pa(uint32_t da, size_t size,
+				   struct remoteproc_context *ctx)
+{
+	uint32_t param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					       TEE_PARAM_TYPE_VALUE_INPUT,
+					       TEE_PARAM_TYPE_VALUE_INPUT,
+					       TEE_PARAM_TYPE_VALUE_OUTPUT);
+	TEE_Param params[TEE_NUM_PARAMS] = { };
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	/*
+	 * The ELF file contains remote processor device addresses, that refer
+	 * to the remote processor memory space.
+	 * A translation is needed to get the corresponding physical address.
+	 */
+
+	params[0].value.a = ctx->fw_id;
+	params[1].value.a = da;
+	params[2].value.a = size;
+
+	res = TEE_InvokeTACommand(pta_session, TEE_TIMEOUT_INFINITE,
+				  PTA_RPROC_FIRMWARE_DA_TO_PA,
+				  param_types, params, NULL);
+	if (res != TEE_SUCCESS) {
+		EMSG("Failed to translate device address %#"PRIx32, da);
+		return 0;
+	}
+
+	return (paddr_t)reg_pair_to_64(params[3].value.b, params[3].value.a);
+}
+
+static TEE_Result remoteproc_parse_rsc_table(struct remoteproc_context *ctx,
+					     uint8_t *fw_img, size_t fw_img_sz,
+					     paddr_t *rsc_pa,
+					     size_t *rsc_size)
+{
+	uint32_t da = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	Elf32_Word size = 0;
+
+	res = e32_parser_find_rsc_table(fw_img, fw_img_sz, &da, &size);
+	if (res)
+		goto err;
```

Comment:
if no ressource table (`TEE_ERROR_NO_DATA`), is that an error?

Reply:
Ok, makes sense. In that case you should update the trace message accordingly:
`IMSG("Ignore empty resource table section");` in _ta/remoteproc/elf_parser.c_ line 145.

Reply:
What about `IMSG("Warning: No resource table found");` Print an error here seems too strong